### DISCUSSION TITLE
audio/video badges displayed with uosc even if stream has only one audio/video track?

### DISCUSSION THREAD

**Discussion by xfzv** [christoph-heinrich/mpv-quality-menu#32] [Q&A] [answered]
I'm trying to have audio/video badges displayed with `uosc` only if current file (local or stream) has multiple audio/video tracks.

The following works as expected with local files:

```
controls=<!stream,has_many_audio>audio,<!stream,has_many_video>video
```

However, I'm having some trouble for stream files using the example from the README:

https://github.com/christoph-heinrich/mpv-quality-menu/blob/9bb4d87681b9765a8035158c9076f4a37b6f7b07/README.md?plain=1#L48-L49

It works fine for streams that have multiple audio/video tracks. However, the badges are still displayed for streams that have only one video/audio quality available, like https://www.twitch.tv/videos/1890295860

![1](https://github.com/christoph-heinrich/mpv-quality-menu/assets/78810647/360b2749-8120-41d3-92ed-f5294fb89189)

```console
yt-dlp -F "https://www.twitch.tv/videos/1890295860"

[twitch:vod] Extracting URL: https://www.twitch.tv/videos/1890295860
[twitch:vod] 1890295860: Downloading stream metadata GraphQL
[twitch:vod] 1890295860: Downloading video access token GraphQL
[twitch:vod] 1890295860: Downloading m3u8 information
[twitch:vod] 1890295860: Downloading storyboard metadata JSON
[info] v1890295860: Downloading subtitles: rechat
[MetadataParser] Parsed meta_title from '%(title)s': 'E SI RITORNA  AL 1998'
[MetadataParser] Parsed meta_artist from '%(channel)s': 'NA'
[SponsorBlock] SponsorBlock is not supported for TwitchVod
[info] Available formats for v1890295860:
ID         EXT   RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC      ACODEC     ABR MORE INFO
────────────────────────────────────────────────────────────────────────────────────────────────
sb1        mhtml 110x62       0 │                  mhtml │ images                     storyboard
sb0        mhtml 220x124      0 │                  mhtml │ images                     storyboard
Audio_Only mp4   audio only     │ ~ 47.46MiB  189k m3u8  │ audio only  mp4a.40.2 189k
480p50     mp4   852x480     34 │ ~374.13MiB 1489k m3u8  │ avc1.4D001F mp4a.40.2      Source
```

The following works (audio/video badges aren't displayed for streams that have only one audio/video track):

```
<stream,has_many_audio,has_audio>
<stream,has_many_video,video>

# or

<stream,has_many_audio>
<stream,has_many_video>
```

but then the audio badge is not displayed for streams that have multiple audio tracks, such as https://yewtu.be/watch?v=z39UHmVaDxk 

```console
yt -F "https://yewtu.be/watch?v=z39UHmVaDxk"
[..]
[info] Available formats for z39UHmVaDxk:
ID  EXT   RESOLUTION FPS CH │   FILESIZE   TBR PROTO │ VCODEC          VBR ACODEC      ABR ASR MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
[..]
599 m4a   audio only      2 │    1.84MiB   31k https │ audio only          mp4a.40.5   31k 22k [en] ultralow, m4a_dash
600 webm  audio only      2 │    2.11MiB   35k https │ audio only          opus        35k 48k [en] ultralow, webm_dash
139 m4a   audio only      2 │    2.91MiB   49k https │ audio only          mp4a.40.5   49k 22k [en] low, m4a_dash
249 webm  audio only      2 │    3.11MiB   52k https │ audio only          opus        52k 48k [en] low, webm_dash
250 webm  audio only      2 │    4.08MiB   68k https │ audio only          opus        68k 48k [en] low, webm_dash
140 m4a   audio only      2 │    7.72MiB  129k https │ audio only          mp4a.40.2  129k 44k [en] medium, m4a_dash
251 webm  audio only      2 │    7.89MiB  132k https │ audio only          opus       132k 48k [en] medium, webm_dash
[..]
18  mp4   640x360     24  2 │   43.14MiB  724k https │ avc1.42001E         mp4a.40.2       44k [en] 360p
605 mp4   640x360     24    │ ~ 41.48MiB  680k m3u8  │ vp09.00.21.08  680k video only
243 webm  640x360     24    │   25.12MiB  422k https │ vp09.00.21.08  422k video only          360p, webm_dash
397 mp4   854x480     24    │   31.33MiB  526k https │ av01.0.04M.08  526k video only          480p, mp4_dash
231 mp4   854x480     24    │ ~ 73.97MiB 1212k m3u8  │ avc1.4D401E   1212k video only
135 mp4   854x480     24    │   39.30MiB  660k https │ avc1.4D401E    660k video only          480p, mp4_dash
606 mp4   854x480     24    │ ~ 69.98MiB 1147k m3u8  │ vp09.00.30.08 1147k video only
244 webm  854x480     24    │   45.15MiB  758k https │ vp09.00.30.08  758k video only          480p, webm_dash
22  mp4   1280x720    24  2 │ ≈ 80.09MiB 1312k https │ avc1.64001F         mp4a.40.2       44k [en] 720p        
```

Any idea?

**Comment by christoph-heinrich** [comment, marked as answer]
When streaming via ytdl mpv itself has only one track (one for video and one for audio), that's why using `has_many_*` doesn't work.

It would be possible to use [`all_formats=yes`](https://mpv.io/manual/master/#options-all-formats) and then have all the formats as tracks in mpv, in which case quality-menu wouldn't be needed anymore. However that comes with it's own set of [problems](https://github.com/mpv-player/mpv/issues/10394). Although as long as you don't use `*d-queue-enable` it might work fine.
But even if it works fine, the information won't be displayed as nicely as with quality-menu :P

I ended up bothered enough by the controls not behaving perfectly that I now use
```
<has_many_video>video,
<has_many_audio>command:audiotrack:script-binding uosc/audio#audio>1?Audio,
<stream>command:stream:script-binding uosc/stream-quality#@vformats>1?Stream quality,
```
instead of what's in the readme. It's not 100% perfect either (can't see the number of audio formats), but it's less bothersome then the alternative.

**Reply by xfzv** [reply]
> When streaming via ytdl mpv itself has only one track (one for video and one for audio), that's why using `has_many_*` doesn't work.

I came to this conclusion too after reading `uosc` `controls` syntax again and again, to the point where I was almost sure the issue wasn't me not understanding it. :disguised_face:

> I ended up bothered enough by the controls not behaving perfectly that I now use
> 
> ```
> <has_many_video>video,
> <has_many_audio>command:audiotrack:script-binding uosc/audio#audio>1?Audio,
> <stream>command:stream:script-binding uosc/stream-quality#@vformats>1?Stream quality,
> ```
> 
> instead of what's in the readme. It's not 100% perfect either (can't see the number of audio formats), but it's less bothersome then the alternative.

Oh, so you don't use `quality-menu` anymore? 

Correct me if I'm wrong but now that I think about it, from a `uosc` user perspective, the only bonus features from `quality-menu` are the built-in reload feature and the separated audio/video buttons, while "Stream quality" from `uosc` only offers one button for video.

If so, I get why you're now using this. `quality-menu` is still a must-have for non-`uosc` users though.

**Reply by christoph-heinrich** [reply]
When `quality-menu` is used together with `uosc`, it [overwrites](https://github.com/christoph-heinrich/mpv-quality-menu/blob/9bb4d87681b9765a8035158c9076f4a37b6f7b07/quality-menu.lua#L1257-L1263) the `stream-quality` menu of uosc.
I couldn't live without dynamically fetched formats and neatly displayed information about those formats, so `quality-menu` is a must for me :wink:

**Reply by xfzv** [reply]
> When `quality-menu` is used together with `uosc`, it [overwrites](https://github.com/christoph-heinrich/mpv-quality-menu/blob/9bb4d87681b9765a8035158c9076f4a37b6f7b07/quality-menu.lua#L1257-L1263) the `stream-quality` menu of uosc. I couldn't live without dynamically fetched formats and neatly displayed information about those formats, so `quality-menu` is a must for me :wink: 

Ha, I didn't realize that. I've just tried `stream-quality` with `quality-menu` disabled and I much prefer `quality-menu` too, of course.

I've played around with your `controls` settings for audio / video and they're nice too. Not having the audio button displayed at all for streams feels more consistent.

I could use them or stick with mine, I almost exclusively use keybindings and never really switch audio quality anyway:

```
a    script-binding quality_menu/audio_formats_toggle    #! Audio
v    script-binding quality_menu/video_formats_toggle    #! Video > Formats
F5   script-binding quality_menu/reload                  #! Video > Reload
```
